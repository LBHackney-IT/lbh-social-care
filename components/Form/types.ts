/* eslint-disable @typescript-eslint/no-explicit-any */

import { InputHTMLAttributes } from 'react';
import { RegisterOptions } from 'react-hook-form';
import TextInput from './TextInput/TextInput';

type LabelSize = 's' | 'm' | 'l' | 'xl';
type Width = 30 | 20 | 10 | 5 | 4 | 3 | 2;
type GovGrid =
  | 'full'
  | 'half'
  | 'one-half'
  | 'one-third'
  | 'two-thirds'
  | 'one-quarter'
  | 'three-quarters';

export type Option = string | { text: string; value: string };

type FuncOption = (arg0: Record<string, string>) => Option[] | null;

export type Options = FuncOption | Option[];

export interface GenericField {
  name: string;
  label?: string;
  rules?: RegisterOptions;
  register?: (arg0: RegisterOptions) => any;
  error?: { message?: string };
  required?: boolean | string;
  hint?: string;
  labelSize?: LabelSize;
  conditionalRender?: (arg0: Record<string, any>) => boolean;
  isMulti?: boolean;
  isMultiInit?: boolean;
  isMultiTrigger?: string;
}

export interface Checkbox extends GenericField {
  options?: Option[];
}

interface StepCheckbox extends Omit<Checkbox, 'options'> {
  component: 'Checkbox';
  options?: Options;
}

type TextInputType = GenericField & InputHTMLAttributes<HTMLInputElement>;

export interface TextInput extends TextInputType {
  inputClassName?: string;
  width?: Width;
}

interface StepTextInput extends TextInput {
  component: 'TextInput';
}

export type TextInputNoType = Omit<TextInput, 'type'>;

interface StepTextInputNoType extends TextInputNoType {
  component: 'NumberInput' | 'DatePicker' | 'PhoneInput' | 'EmailInput';
}

type TextAreaType = GenericField & InputHTMLAttributes<HTMLTextAreaElement>;

export interface TextArea extends TextAreaType {
  width?: Width;
  rows?: number;
}

interface StepTextArea extends TextArea {
  component: 'TextArea';
}

type SelectType = GenericField &
  Omit<InputHTMLAttributes<HTMLSelectElement>, 'onChange'>;

export interface Select extends SelectType {
  options: Option[];
  placeHolder?: string;
  ignoreValue?: boolean;
  isUnselectable?: boolean;
  width?: Width;
  govGrid?: GovGrid;
  onChange?: (arg0: string) => string | void;
}

interface StepSelect extends Omit<Select, 'options'> {
  component: 'Select';
  options: Options;
}

type RadiosType = GenericField & InputHTMLAttributes<HTMLInputElement>;

export interface Radios extends RadiosType {
  options?: Option[];
  isRadiosInline?: boolean;
  children?: React.ReactNode;
}

interface StepRadios extends Omit<Radios, 'options'> {
  component: 'Radios';
  options?: Options;
}

export interface ObjectInput extends GenericField {
  isMulti?: boolean;
  isInline?: boolean;
  summaryInline?: boolean;
  components: Array<formBasicComponentStep>;
}

interface StepObjectInput extends ObjectInput {
  component: 'ObjectInput';
}

export interface AddressLookup extends GenericField {
  supportManualEntry?: boolean;
}

interface StepAddressLookup extends AddressLookup {
  component: 'AddressLookup';
}

export interface DateInput extends GenericField {
  format?: 'US' | 'EU';
}

interface StepDateInput extends DateInput {
  component: 'DateInput';
}

type formBasicComponentStep =
  | StepTextInput
  | StepTextInputNoType
  | StepTextArea
  | StepCheckbox
  | StepSelect
  | StepRadios
  | StepAddressLookup
  | StepDateInput;

export type FormComponentStep =
  | formBasicComponentStep
  | StepObjectInput
  | JSX.Element;

export interface FormStep {
  id: string;
  title: string;
  components: Array<FormComponentStep>;
  isMulti?: boolean;
  conditionalRender?: (arg0: Record<string, unknown>) => boolean;
}

export interface FormConfig {
  title: string;
  path: string;
  successMessage?: string;
  steps: Array<FormStep>;
}
