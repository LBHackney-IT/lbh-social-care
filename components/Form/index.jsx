import AddressLookup from './AddressLookup/AddressLookup';
import Autocomplete from './Autocomplete/Autocomplete';
import Checkbox from './Checkbox/Checkbox';
import DateInput from './DateInput/DateInput';
import DatePicker from './DatePicker/DatePicker';
import Radios from './Radios/Radios';
import Select from './Select/Select';
import TextInput from './TextInput/TextInput';
import TextArea from './TextArea/TextArea';
import EmailInput from './EmailInput/EmailInput';
import NumberInput from './NumberInput/NumberInput';
import PhoneInput from './PhoneInput/PhoneInput';
import ObjectInput from './ObjectInput/ObjectInput';

export {
  AddressLookup,
  Autocomplete,
  Checkbox,
  DateInput,
  DatePicker,
  Radios,
  Select,
  TextInput,
  TextArea,
  EmailInput,
  NumberInput,
  PhoneInput,
  ObjectInput,
};
