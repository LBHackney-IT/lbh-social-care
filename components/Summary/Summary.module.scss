.sectionTitle {
  display: flex;
  align-items: center;
  a {
    margin-left: 1rem;
  }
}

.summaryControls {
  display: flex;
  justify-content: flex-end;
}

.formattedTextArea {
  white-space: pre-line;
}
