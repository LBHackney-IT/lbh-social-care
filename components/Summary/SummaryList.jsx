import PropTypes from 'prop-types';

const SummaryList = ({ list }) => (
  <dl className="govuk-summary-list">
    {list &&
      list.map(({ key, title, value, href }) => (
        <div key={key} className="govuk-summary-list__row">
          <dt className="govuk-summary-list__key">{title}</dt>
          <dd className="govuk-summary-list__value">{value}</dd>
          {href && (
            <dd className="govuk-summary-list__actions">
              <a className="govuk-link" href="#">
                Change
                <span className="govuk-visually-hidden"> {title}</span>
              </a>
            </dd>
          )}
        </div>
      ))}
  </dl>
);

SummaryList.propTypes = {
  list: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.node.isRequired,
      href: PropTypes.string,
    }).isRequired
  ),
};

export default SummaryList;
