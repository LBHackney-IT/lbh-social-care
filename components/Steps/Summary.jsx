import { useState } from 'react';
import PropTypes from 'prop-types';
import Router from 'next/router';

import { Button } from 'components/Form';
import { deleteData } from 'utils/saveData';
import Summary from 'components/Summary/Summary';
import ErrorSummary from 'components/ErrorSummary/ErrorSummary';
import { filterDataOnCondition } from 'utils/steps';

const SummaryStep = ({ formData, formSteps, formPath, onFormSubmit }) => {
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [hasError, setHasError] = useState(false);
  const onSubmit = async () => {
    setIsSubmitting(true);
    setHasError(false);
    try {
      const data = await onFormSubmit(
        filterDataOnCondition(formSteps, formData)
      );
      deleteData(formPath);
      Router.replace(
        `${formPath}confirmation${data?.ref ? `?ref=${data.ref}` : ''}`
      );
    } catch {
      setHasError(true);
    }
    setIsSubmitting(false);
  };
  if (!formSteps) return null;
  return (
    <div>
      <Summary
        formData={formData}
        formPath={formPath}
        formSteps={formSteps}
        canEdit
      />
      <Button
        className="govuk-button"
        label="Submit"
        onClick={onSubmit}
        disabled={isSubmitting}
      />
      {hasError && (
        <ErrorSummary
          title="Unfortunately there was a problem with your submission."
          body="Please try again."
        />
      )}
    </div>
  );
};

SummaryStep.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  formSteps: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  formPath: PropTypes.string.isRequired,
  onFormSubmit: PropTypes.func.isRequired,
};

export default SummaryStep;
