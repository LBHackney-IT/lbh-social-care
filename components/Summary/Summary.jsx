import { useState } from 'react';
import cx from 'classnames';
import PropTypes from 'prop-types';
import Link from 'next/link';

import SummaryList from 'components/Summary/SummaryList';
import { filterStepsOnCondition, filterDataOnCondition } from 'utils/steps';
import { formatData, getSectionObject } from 'utils/summary';
import { setValues } from 'utils/objects';

import styles from './Summary.module.scss';

const SummaryMultiSection = ({
  formData,
  id,
  title,
  canEdit,
  formPath,
  ...props
}) => (
  <>
    {formData[id].map((data, key) => (
      <SummarySection
        {...props}
        canEdit={canEdit}
        formData={data}
        formPath={formPath}
        title={`${title} - ${key + 1}`}
        key={`${id}/${key}`}
        id={`${id}/${key + 1}`}
      />
    ))}
    {canEdit && (
      <p className="govuk-!-margin-bottom-7">
        <Link
          href={`${formPath}${id}/${formData[id].length + 1}?fromSummary=true`}
          as={`${formPath}${id}/${formData[id].length + 1}?fromSummary=true`}
        >
          <a className="govuk-link">Add Another {title}</a>
        </Link>
      </p>
    )}
  </>
);

SummaryMultiSection.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  formPath: PropTypes.string.isRequired,
  canEdit: PropTypes.bool,
  additionalMetadata: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      type: PropTypes.string,
      href: PropTypes.string,
    })
  ),
};

export const SummarySection = ({
  formData,
  id,
  title,
  components,
  formPath,
  canEdit,
  collapsedSection,
  toggleCollapsed,
  isSummaryCollapsable,
  additionalMetadata,
}) => {
  const isCollapsed = collapsedSection[id];

  return (
    <div className="govuk-!-margin-bottom-7">
      <div className="lbh-table-header">
        <div className={styles.sectionTitle}>
          <h3 className="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0">
            {title.toUpperCase()}
          </h3>
          {canEdit && (
            <Link
              href={`${formPath}${id}?fromSummary=true`}
              as={`${formPath}${id}?fromSummary=true`}
            >
              <a className="govuk-link">Edit</a>
            </Link>
          )}
        </div>
        {isSummaryCollapsable && (
          <button
            className="govuk-link govuk-link--underline"
            onClick={(e) => {
              e.preventDefault();
              toggleCollapsed(id);
            }}
          >
            {isCollapsed ? 'Expand view' : 'Collapse view'}
          </button>
        )}
      </div>
      <hr className="govuk-divider" />
      {!isCollapsed && (
        <SummaryList
          list={[
            ...components
              .filter(({ name }) => formData[name])
              .map((data) => formatData(data, formData)),
            ...(additionalMetadata || []),
          ]}
        />
      )}
    </div>
  );
};

SummarySection.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  id: PropTypes.string.isRequired,
  title: PropTypes.string.isRequired,
  components: PropTypes.arrayOf(
    PropTypes.shape({
      name: PropTypes.string,
    }).isRequired
  ).isRequired,
  formPath: PropTypes.string.isRequired,
  canEdit: PropTypes.bool,
  collapsedSection: PropTypes.object.isRequired,
  toggleCollapsed: PropTypes.func.isRequired,
  isSummaryCollapsable: PropTypes.bool,
  additionalMetadata: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      type: PropTypes.string,
      href: PropTypes.string,
    })
  ),
};

const Summary = ({
  formData,
  formSteps,
  formPath,
  canEdit,
  isSummaryCollapsable,
  additionalMetadata,
}) => {
  const [collapsedSection, setCollapsedSection] = useState(
    getSectionObject(formSteps, formData, false)
  );
  const hasCollapsed = Object.values(collapsedSection).find(Boolean);
  return (
    <>
      {isSummaryCollapsable && (
        <div className={styles.summaryControls}>
          <button
            className={cx(
              'govuk-link',
              'govuk-link--underline',
              'govuk-!-font-weight-bold'
            )}
            onClick={(e) => {
              e.preventDefault();
              setCollapsedSection(setValues(collapsedSection, !hasCollapsed));
            }}
          >
            {hasCollapsed ? 'Expand all' : 'Collapse all'}
          </button>
        </div>
      )}
      {filterStepsOnCondition(formSteps, formData).map((section) => {
        const props = {
          key: section.id,
          formData: filterDataOnCondition(formSteps, formData),
          formPath: formPath,
          canEdit: canEdit,
          collapsedSection,
          toggleCollapsed: (id) =>
            setCollapsedSection({
              ...collapsedSection,
              [id]: !collapsedSection[id],
            }),
          isSummaryCollapsable,
          ...section,
          additionalMetadata,
        };
        return section.isMulti ? (
          <SummaryMultiSection {...props} />
        ) : (
          <SummarySection {...props} />
        );
      })}
    </>
  );
};

Summary.propTypes = {
  formData: PropTypes.shape({}).isRequired,
  formSteps: PropTypes.array.isRequired,
  formPath: PropTypes.string.isRequired,
  canEdit: PropTypes.bool,
  isSummaryCollapsable: PropTypes.bool,
  additionalMetadata: PropTypes.arrayOf(
    PropTypes.shape({
      key: PropTypes.string.isRequired,
      title: PropTypes.string.isRequired,
      value: PropTypes.string.isRequired,
      type: PropTypes.string,
      href: PropTypes.string,
    })
  ),
};

export default Summary;
