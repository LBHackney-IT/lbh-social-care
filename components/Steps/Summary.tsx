import React, { useState } from 'react';
import Router from 'next/router';

import Button from 'components/Button/Button';
import { deleteData } from 'utils/saveData';
import Summary from 'components/Summary/Summary';
import ErrorSummary from 'components/ErrorSummary/ErrorSummary';
import { filterDataOnCondition } from 'utils/steps';
import { sanitiseObject } from 'utils/objects';
import { FormStep } from 'components/Form/types';

interface Props {
  formData: Record<string, unknown>;
  formSteps: FormStep[];
  formPath: string;
  onFormSubmit: (filteredData: unknown) => Promise<{ ref?: string } | void>;
  isSummaryCollapsable?: boolean;
}

const SummaryStep = ({
  formData,
  formSteps,
  formPath,
  onFormSubmit,
  isSummaryCollapsable,
}: Props): React.ReactElement => {
  const [isSubmitting, setIsSubmitting] = useState(false);
  const [hasError, setHasError] = useState(false);
  const onSubmit = async () => {
    setIsSubmitting(true);
    setHasError(false);
    try {
      const data = await onFormSubmit(
        //@ts-ignore TODO fix me
        filterDataOnCondition(formSteps, sanitiseObject(formData))
      );
      deleteData(formPath);
      Router.replace(
        `${formPath}confirmation${data && data.ref ? `?ref=${data.ref}` : ''}`
      );
    } catch {
      setHasError(true);
    }
    setIsSubmitting(false);
  };
  return (
    <div>
      <h2 className="lbh-heading-h1 govuk-!-margin-bottom-7">
        {`Review details`}
      </h2>
      <Summary
        formData={formData}
        formPath={formPath}
        formSteps={formSteps}
        canEdit
        isSummaryCollapsable={isSummaryCollapsable}
      />
      <div style={{ display: 'flex', justifyContent: 'space-between' }}>
        <Button
          wideButton
          className="govuk-button"
          label="Cancel"
          isSecondary
          onClick={() => (window.location.href = '/')}
        />
        <Button
          wideButton
          className="govuk-button"
          label="Submit"
          onClick={onSubmit}
          disabled={isSubmitting}
        />
      </div>
      {hasError && (
        <ErrorSummary
          title="There was a problem with your submission."
          body="Please try again."
        />
      )}
    </div>
  );
};

export default SummaryStep;
