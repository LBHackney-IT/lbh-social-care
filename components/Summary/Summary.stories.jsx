import Summary from './Summary';

export default {
  title: 'Summary',
  component: Summary,
};

const Template = (args) => <Summary {...args} />;

export const Default = Template.bind({});
Default.args = {
  formData: {
    foo: '123',
    bar: 'asd',
    asd: '',
    qwe: 'yo',
    foobar: 'asd',
    textarea_verbose:
      'Lorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.\nLorem ipsum dolor sit amet, consectetur adipiscing elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. Ut enim ad minim veniam, quis nostrud exercitation ullamco laboris nisi ut aliquip ex ea commodo consequat. Duis aute irure dolor in reprehenderit in voluptate velit esse cillum dolore eu fugiat nulla pariatur. Excepteur sint occaecat cupidatat non proident, sunt in culpa qui officia deserunt mollit anim id est laborum.',
    textarea_succinct: 'asd\nasd',
  },
  formPath: '/form/foo/',
  formSteps: [
    {
      id: 'first-step',
      title: 'First Step',
      components: [
        { component: 'TextInput', label: 'I am foo component', name: 'foo' },
        { component: 'TextInput', label: 'I am bar component', name: 'bar' },
        { component: 'TextInput', label: 'I am asd component', name: 'asd' },
      ],
    },
    {
      id: 'second-step',
      title: 'Second Step',
      components: [
        {
          component: 'TextInput',
          label: 'I am foobar component',
          name: 'foobar',
        },
      ],
    },
    {
      id: 'third-step',
      title: 'TextArea Step',
      components: [
        {
          component: 'TextArea',
          label: 'I am a verbose textarea',
          name: 'textarea_verbose',
        },
        {
          component: 'TextArea',
          label: 'I am a succinct textarea',
          name: 'textarea_succinct',
        },
      ],
    },
  ],
  onFormSubmit: () => alert('submit'),
};
