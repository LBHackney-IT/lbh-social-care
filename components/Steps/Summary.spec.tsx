import { render, fireEvent, waitFor } from '@testing-library/react';
import Summary from './Summary';

import { FormStep } from 'components/Form/types';

describe('Summary component', () => {
  const props = {
    formData: {
      bar_input: 'foo',
    },
    formSteps: [
      {
        components: [
          {
            component: 'TextInput',
            name: 'bar_input',
            width: 30,
            label: 'Foo',
          },
        ],
        id: 'foo-bar',
        title: 'Foo Bar',
      },
    ] as FormStep[],
    onFormSubmit: jest.fn(),
    formPath: 'foo/bar-foo',
  };
  it('should render properly', async () => {
    const { asFragment, getByRole, getByText } = render(<Summary {...props} />);
    const submitBtn = getByRole('button', { name: 'Submit' });
    expect(asFragment()).toMatchSnapshot();
    expect(getByText('Foo')).toBeInTheDocument();
    fireEvent.click(submitBtn);
    await waitFor(() => expect(props.onFormSubmit).toHaveBeenCalled());
  });
});
