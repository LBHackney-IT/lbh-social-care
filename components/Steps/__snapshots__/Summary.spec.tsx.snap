// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Summary component should render properly 1`] = `
<DocumentFragment>
  <div>
    <div
      class="lbh-table-header"
    >
      <h2
        class="govuk-fieldset__legend--l gov-weight-lighter "
      >
        Review details
      </h2>
    </div>
    <div
      class="govuk-!-margin-bottom-7"
    >
      <div
        class="lbh-table-header"
      >
        <div
          class="sectionTitle"
        >
          <h3
            class="govuk-heading-m govuk-custom-text-color govuk-!-margin-bottom-0"
          >
            FOO BAR
          </h3>
          <a
            class="govuk-link"
            href="/foo/bar-foofoo-bar?fromSummary=true"
          >
            Edit
          </a>
        </div>
      </div>
      <hr
        class="govuk-divider"
      />
      <dl
        class="govuk-summary-list"
      >
        <div
          class="govuk-summary-list__row"
        >
          <dt
            class="govuk-summary-list__key"
          >
            Foo
          </dt>
          <dd
            class="govuk-summary-list__value"
          >
            foo
          </dd>
        </div>
      </dl>
    </div>
    <div
      style="display: flex; justify-content: space-between;"
    >
      <button
        class="lbh-button govuk-button lbh-button--secondary govuk-button"
        data-module="govuk-button"
        style="min-width: 200px;"
      >
        Cancel
      </button>
      <button
        class="lbh-button govuk-button govuk-button"
        data-module="govuk-button"
        style="min-width: 200px;"
      >
        Submit
      </button>
    </div>
  </div>
</DocumentFragment>
`;
